### ISSUE TITLE
sample code for making external http calls  from extensions

### ISSUE THREAD

**Issue by sriramcms** [solo-io/wasm#105]
Hi:
Do we have any tutorials , sample code on how to make external http calls from the extensions code.
For example i was able to add headers using the assemblyscript extension  code.  How can i make external http calls etc

Thanks

**Comment by yuval-k** [comment]
hi, you can see a c++ example here:

https://github.com/envoyproxy/envoy-wasm/blob/dfb104874b3babc7e676aaa0199f2926410c0671/test/extensions/filters/http/wasm/test_data/async_call_cpp.cc#L15

**Comment by yuval-k** [comment]
as for assembly script, there's a PR in progress that will also has an example it in

**Comment by sriramcms** [comment]
Thanks.  I progressed to some extent -  I am always ending up with a BAD_ARGUMENT HTTP call wasm return value --  I have pasted details of the assemblyscript code used and also the envoy wasm context.cc which is throwing this error. Unable to figure out what is missing.  

*********  sample  **********
I have a standard istio bookinfo sample deployed.  Added a petstore pod additionally where i inject the below wasm extension .   It gets invoked and behaves as described.  So when i make a POST call to petstore addPets -  the filter gets invoked and within that I do a HttpCall to "details" service in the bookinfo sample (just to see if i can make a external http call) 

Do i need to also add the cluster referenced in the httpcall to the envoyfilter as described here ? -- https://github.com/istio/istio/wiki/EnvoyFilter-Samples

*******************************************
Env = istio 1.5.1 , Kubernetes 1.17.3 , minikube
Using assemblyscript = 0.9.4 and solo-io / proxy-runtime 0.1.6 
******************************************
add_header istio wasm extension  using the webassemblyhub to publish and then deploy using wasme 
`
class AddHeader extends Context {
   onRequestBody(blen: size_t, estrm: bool): FilterDataStatusValues {

   let headers : Headers = []
    let pair : HeaderPair = new HeaderPair()
    pair.key = String.UTF8.encode(":method");
    pair.value = String.UTF8.encode("GET");
    headers.push(pair);

    pair = new HeaderPair()
    pair.key = String.UTF8.encode(":path");
    pair.value = String.UTF8.encode("/details/123");
    headers.push(pair);

    pair = new HeaderPair()
    pair.key = String.UTF8.encode(":host"); < tried "Host"  also 
    pair.value = String.UTF8.encode("details:9080");
    headers.push(pair);

    pair = new HeaderPair()
    pair.key = String.UTF8.encode(":authority");
    pair.value = String.UTF8.encode("details.default.svc.cluster.local");  <-  tried "details" too  
    headers.push(pair);

    // Always end up in a BAD_ARGUMENT = 2 return error 
    let resVal = root_context.httpCall("details", headers, new ArrayBuffer(0), [], 5000, this.context_id, authCB)
    }
}

export function authCB(origin_context_id:u32, headers: u32, body_size: usize, trailers: u32): void {
    log(LogLevelValues.debug, "Inside AUTH CB ");
}
**************************

Code which is giving  BAD_ARGUMENT 

envoy-wasm/blob/master/source/extensions/common/wasm/context.cc

  auto cluster_string = std::string(cluster);
  if (clusterManager().get(cluster_string) == nullptr) {
    **return WasmResult::BadArgument;**
  }

  Http::RequestMessagePtr message(
      new Http::RequestMessageImpl(buildRequestHeaderMapFromPairs(request_headers)));

  // Check that we were provided certain headers.
  if (message->headers().Path() == nullptr || message->headers().Method() == nullptr ||
      message->headers().Host() == nullptr) {
    **return WasmResult::BadArgument;**
  }
  






`

**Comment by yuval-k** [comment]
the first argument to `httpCall` is the cluster name, look at the enovy config dump to see the list of clusters

**Comment by sriramcms** [comment]
Thanks it worked.  

Ran the http://127.0.0.1:15000/config_dump to get the actual cluster name.  
It turns out to be "outbound|9080||details.default.svc.cluster.local". 

So the above code  with method , path , authority and the above one for cluster.  I was able to make the httpCall and get the response.  Printed the response within the callback  "authCB" defined above
***************************************************************************************

A few more questions -- 

I was able to get the response body via this in the callback function 

 let bytes =  get_buffer_bytes(BufferTypeValues.HttpCallResponseBody,0,body_size);
 String.UTF8.decode(bytes);

The headers: u32 as defined in the callback function params -- how do i use that to get the headers content ?

On other observation which is just cosmetic is that it prints this 
**" wasm log add_header: onHttpCallResponse: Token 1 not found."  -** 

In /v0.1.6/assembly/runtime.ts - onHttpCallResponse the below statement happens even on a successful token existence 
log(LogLevelValues.error, "onHttpCallResponse: Token " + token.toString() + " not found.");

**Comment by yuval-k** [comment]
good catch - i'll fix that. let me see if i can prepare an example

**Comment by sriramcms** [comment]
Thanks.

Essentially - once i make this external HttpCall -- I would want the data from the  call response header and body to be injected into the original request  via the stream_context.headers.request  for the upstream services to use.  It could be injected security headers etc.

**Comment by yuval-k** [comment]
I added a basic example that you should be able to modify to do this.  See examples/auth

**Comment by RoeeShlomo** [comment]
@yuval-k can you please share  a link to the example? I can't find it

**Comment by sriramcms** [comment]
Thanks.  I followed the flow in the onRequestHeaders by making the external httpCall -  I am observing this  when i print the log times  

For example I have a petstore app which prints the http headers for any  call  with the timestamp (including milliseconds).The wasme filter  is injected into the petstore app. 

1. The sequence of log outputs based on timestamp -  wasme filters onRequestHeader , OnRequestBody   calls  followed with the petstore call and then the onResponseHeaders followed by the  original httpCall  -- callback function.
2.   Whatever is add to the headers -- stream_context.header.request.add --  shows up in the petstore app dump headers -- any additions before the httpCall 
3.   What i want to do is the take the response output of the httpCall obtained via the callback functions --  and add them to the header of the original request 
-- since the request already is in the petstore app before the httpCall call back functions is invoked -- how can i control this sequence -- the callback function invocation is async and happens after

Anything to wait for the httpCall callback before forwarding the request to the petstore app container ?

**Comment by yuval-k** [comment]
no need to wait; you can add the headers in the callback, after setEffectiveContext is called.
setEffectiveContext() restores the context to the request context. 
then call `stream_context.headers.request.add`
and finally call `continue_request`

**Comment by sriramcms** [comment]
This is the sequence and  the issue i am facing
**************************************************

A. 2020-04-22 11:54:55.610 - Inside onRequestHeaders and HttpCall response done from that method

B. 2020/04/22 11:54:55.610927 -  Inside actual Petstore app iterating headers 

C. 2020/04/22 11:54:55.612   -- Inside onResponseHeaders 

D. 2020-04-22 11:54:55.653 - Inside HttpCall Callback Invocation 

If you see long after the reponse (C) is sent  at 55.653 the Callback (D) gets invokes as if in the background.  The headers added in C is never seen in the Petstore app when we print the request headers. 

Anything i add to stream_context before the httpCall is seen in the petstore app -- it is just that the changes in call back ate never seen. 

// Code in onrequest headers 
 onRequestHeaders(a: u32) {
   const root_context = this.root_context;
   let resVal = root_context.httpCall("outbound|9080||details.default.svc.cluster.local", headers, new ArrayBuffer(0), [], 5000, this, authCB)
 }

// Code in Callback 
export function authCB(origin_context: Context, headers: u32, body_size: usize, trailers: u32): void {
    let context = origin_context as AddHeader;
    context.setEffectiveContext();
    **stream_context.headers.request.add("hello22AuthCB", "world!22AuthCB");**  <-- never seen by petstore 
   // print stream_context.headers.http_callback.get_headers();
   continue_request(); 
}

**Comment by yuval-k** [comment]
just tested you use case and you are right, with istio 1.5, the request headers are not available in the callback. this is already fixed in upstream envoy-wasm, just not in the version you are testing.

i created a pr with a working example here (test with `istio/proxyv2:1.6.0-alpha.1`): https://github.com/solo-io/proxy-runtime/pull/16
unfortunately in that envoy version some constants have changed, so it is not backwards compatible. as such, I will have to do some thinking prior to merging this

**Comment by sriramcms** [comment]
okay - let me change my istio version to  https://github.com/istio/istio/tree/1.6.0-alpha.1 and try to run the same. I will use the example you gave and report.

**Comment by yuval-k** [comment]
@sriramcms do note that you need to build with the updated runtime, because of the changes to the constants.
the easiest way to do so, is to git clone it, checkout the branch of the PR, and point to it in your `package.json` like so:
```
"@solo-io/proxy-runtime": "file:/PATH/TO/proxy-assemblyscript"
```
and run `npm install` again before building.

**Comment by sriramcms** [comment]
I built the  wasm filter this way 
package.json 
  "dependencies": {
    "@assemblyscript/loader": "^0.9.4",
    "@solo-io/proxy-runtime": "file:./../"
  },

$HOME/.wasme/bin/wasme build precompiled ./build/optimized.wasm -t webassemblyhub.io/sriramcm/demo-add-header:v0.24
**INFO[0000] adding image to cache...                      filter file=./build/optimized.wasm tag="webassemblyhub.io/sriramcm/demo-add-header:v0.24"**
INFO[0000] tagged image                                  digest="sha256:6726087078f75811efc9f86cc8fc1d9fca5dd72ba7967f22c8c782727be0badf" image="webassemblyhub.io/sriramcm/demo-add-header:v0.24"


**When i apply to the petstore app -- i get this error** 
INFO[0000] cache rolebinding updated                     cache=wasme-cache.wasme image="quay.io/solo-io/wasme:0.0.19"
**Error: image webassemblyhub.io/sriramcm/demo-add-header:v0.24 not supported by istio version 1.6.0-alpha.1**

**Comment by yuval-k** [comment]
`wasme` is being conservative here; essentially in order to prevent mistakes, wasme white lists which api version is compatible with which istio version;

@ilackarms what would be the best course of action here?
probably extend wasme to support istio 1.6 beta1?

**Comment by roee88** [comment]
Can there be a flag for force apply and skip this check? Would be useful for dev and won't require adding any alpha/beta explicitly.

**Comment by sriramcms** [comment]
Instead of using wasme and webasseblyhub -- i create a configmap  with optimized.wasm  and then mounted that as a path via annotation in the petstore app 
      annotations:
        sidecar.istio.io/userVolume: '[{"name":"wasmfilters-dir","configMap": {"name": "petstore-filter"}}]'
        sidecar.istio.io/userVolumeMount: '[{"mountPath":"/var/local/lib/wasm-filters","name":"wasmfilters-dir"}]'      

*************** here is the analysis 
Same sequence of missing the headers set in the HttpCall callback as by that time the call already returns from petstore --  I am suspecting that the httpCall does not wait for the callback and instead immediately returns and the request proceeds to petstore 

My intention is to use the response from the httpCall callback to augment the request headers and then allow it to go to petstore -- looks like i dont have control on this aspect -- any thoughts ?


****************************  Please check the timings of the call sequence** 
**OnrequestHeader ----**
/wasm/context.cc:1098] wasm log myfilter add_header: Response of httpcall 0
2020-04-24T20:38:46.913891Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter 

**OnrequestBody ----**
add_header: Inside onrequestBody 12 New false
2020-04-24T20:38:46.913919Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter add_header: Inside onrequestBody 11 {"id":3,"name":"Tiger","status":"available"}
2020-04-24T20:38:46.913946Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter add_header: Inside onrequestBody 12 New true
[2020-04-24T20:38:46.910Z] "POST /api/pets HTTP/1.1" 200 - "-" "-" 44 45 9 3 "-" "curl/7.58.0" "ac8a5c7d-ed56-91ec-ad8e-2eb0444d03d5" "petstore:8080" "127.0.0.1:8080" inbound|8080|http|petstore.default.svc.cluster.local 127.0.0.1:32844 172.17.0.17:8080 172.17.0.15:51560 - default


**Inside Actual Petstore App**
****************************
2020/04/24 20:38:46.915568 Request: &http.Request{Method:"POST", URL:

2020/04/24 20:38:46.916233 Iterating Headers
2020/04/24 20:38:46.916262 Header field "User-Agent", Value ["curl/7.58.0"]
2020/04/24 20:38:46.916436 Header field "Accept", Value ["*/*"]
2020/04/24 20:38:46.916453 Header field "Content-Type", Value ["application/json"]
2020/04/24 20:38:46.916460 Header field "Content-Length", Value ["44"]
2020/04/24 20:38:46.916470 Header field "X-Request-Id", Value ["ac8a5c7d-ed56-91ec-ad8e-2eb0444d03d5"]
2020/04/24 20:38:46.916478 Header field "Hello22init", Value ["world22init"]
2020/04/24 20:38:46.916485 Header field "X-B3-Spanid", Value ["6cea313e9123ab86"]
2020/04/24 20:38:46.916492 Header field "X-Forwarded-Proto", Value ["http"]
2020/04/24 20:38:46.916499 Header field "Hello32init", Value ["world32init"]
2020/04/24 20:38:46.916507 Header field "X-B3-Traceid", Value ["1c7146e7b3573e8f6cea313e9123ab86"]
2020/04/24 20:38:46.916514 Header field "X-B3-Sampled", Value ["1"]
createPet



**OnresponseHeaders ----**
2020-04-24T20:38:46.917972Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter add_header: Inside onresponse headers
2020-04-24T20:38:47.173170Z     critical        envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1101] wasm log add_header: Unable to set effective context: 2 with result: 2



**OnrequestHeaders HttpCall callback ----**
2020-04-24T20:38:47.173291Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log add_header: Inside AUTH CB MODIFIED 2222222
2020-04-24T20:38:47.173301Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log add_header: Inside AUTH CB Response Call Headers 11
2020-04-24T20:38:47.173333Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log add_header: :status
2020-04-24T20:38:47.173341Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log add_header: 200


For reference the code inside the httpCall callback sets the below headers  which you will find missing in the petstore -- headers dump

    let context = origin_context as AddHeader;
    //context.setEffectiveContext();
    stream_context.headers.request.add("hello22AuthCB", "world!22AuthCB");

**Comment by sriramcms** [comment]
Thanks it worked.   Used the Istio 1.6 alpha 2 and applied envoyfilter directly without going through wasme.    

***************  The sequence works perfectly 
**END OF HttpCall CALLBACK**
2020-04-25T13:52:56.436886Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter add_header: callback continue request

**INSIDE THE PETSTORE APP HEADERS ITERATION**
2020/04/25 13:52:56.437989 Iterating Headers
2020/04/25 13:52:56.438037 Header field "Hello22init", Value ["world22init"]
2020/04/25 13:52:56.438050 Header field "Hello22authcb", Value ["world!22AuthCB"]
2020/04/25 13:52:56.438071 Header field "Content-Type", Value ["application/json"]
2020/04/25 13:52:56.438177 Header field "Added-Header", Value ["authorized"]


**RESPONSE  HEADERS** 
2020-04-25T13:52:56.439660Z     error   envoy wasm      [external/envoy/source/extensions/common/wasm/context.cc:1098] wasm log myfilter add_header: Inside onresponse headers

**Comment by esnible** [comment]
Another request for a `wasme deploy istio` with `--force` flag.

I have verified that if I do `wasme deploy istio` on Istio 1.5.4, then upgrade Istio to 1.6.1, that the ` webassemblyhub.io/$YOUR_USERNAME/add-header:v0.1` example still works on the running bookinfo pods.

It took me a bit longer to verify, but if I delete bookinfo and recreate it leaving the EnvoyFilter CRs untouched, and manually set the `sidecar.istio.io/userVolume` and `sidecar.istio.io/userVolumeMount` annotations, it all still works.

Adding Istio 1.6.x to the white list isn't enough.  I am also using daily builds with the image tag `latest`.

**Comment by esnible** [comment]
Work-around to force `wasme` to believe it is working with Istio 1.5.4 no matter what the version:

```
IMAGE=$(kubectl -n istio-system get deployment istiod -o jsonpath="{.spec.template.spec.containers[0].image}")
ALIAS=docker.io/$YOUR_USERNAME/pilot:1.5.4
docker pull $IMAGE
docker tag $IMAGE $ALIAS
docker push $ALIAS
kubectl -n istio-system patch deployment istiod --type=json --patch="[{\"op\": \"replace\", \"path\": \"/spec/template/spec/containers/0/image\", \"value\": "$ALIAS"}]"
```

**Comment by tomersolomon1** [comment]
It seems another work-around can be achieved by just not specifying any ABI version:
```
{
  "type": "envoy_proxy",
  "abiVersions": [
  ],
  "config": {
    "rootIds": [
      <filter-name>
    ]
  }
}
```
It seems in that case the ABI check is skipped, and you get a log like this:
```
time="2020-06-21T14:43:02Z" level=warning msg="no ABI Version found for image, skipping ABI version check"
```
For more details, this is the relevant code as far as I understand: https://github.com/solo-io/wasme/blob/master/pkg/deploy/istio/istio_provider.go#L132

**Comment by Sodman** [comment]
Looks like the sample code provided by Yuval is working now. Separately the requested Istio force deploy feature has been implemented by https://github.com/solo-io/wasm/pull/179 and is available in wasme v0.0.27+, so I'm going to close this one out.
